utils/logger/logger.go

```diff
@@ -0,0 +1,205 @@
+	// method. If we also set it up here, messages will be written to the console twice. Eventually, when all instances
+	// of l4g have been replaced by this logger, we can move that code to here
+	if config.LogSettings.EnableFile {
+		level := l4g.DEBUG
+		if config.LogSettings.FileLevel == "INFO" {
+			level = l4g.INFO
+		} else if config.LogSettings.FileLevel == "WARN" {
+			level = l4g.WARNING
+		} else if config.LogSettings.FileLevel == "ERROR" {
+			level = l4g.ERROR
+		}
+
+		// create a logger that writes JSON objects to a file, and override our log methods to use it
+		flw := NewJSONFileLogger(level, utils.GetLogFileLocation(config.LogSettings.FileLocation)+".jsonl")
+		debug = flw.Debug
+		info = flw.Info
+		err = flw.Error
+	}
+}
+
+// contextKey lets us add contextual information to log messages
+type contextKey string
+
+func (c contextKey) String() string {
+	return string(c)
+}
+
+const contextKeyUserID contextKey = contextKey("user-id")
+const contextKeyRequestID contextKey = contextKey("request-id")
+
+// any contextKeys added to this array will be serialized in every log message
+var contextKeys = [2]contextKey{contextKeyUserID, contextKeyRequestID}
+
+// WithUserID adds a user id to the specified context. If the returned Context is subsequently passed to a logging
+// method, the user id will automatically be included in the logged message
+func WithUserID(ctx context.Context, userID string) context.Context {
+	return context.WithValue(ctx, contextKeyUserID, userID)
+}
+
+// WithRequestID adds a request id to the specified context. If the returned Context is subsequently passed to a logging
+// method, the request id will automatically be included in the logged message
+func WithRequestID(ctx context.Context, requestID string) context.Context {
+	return context.WithValue(ctx, contextKeyRequestID, requestID)
+}
+
+// extracts known contextKey values from the specified Context and assembles them into the returned map
+func serializeContext(ctx context.Context) map[string]string {
+	serialized := make(map[string]string)
+	for _, key := range contextKeys {
+		value, ok := ctx.Value(key).(string)
+		if ok {
+			serialized[string(key)] = value
+		}
+	}
+	return serialized
+}
+
+// returns the path to the next file up the callstack that has a different name than this file
+// in other words, finds the path to the file that is doing the logging
+// looks a maximum of 10 frames up the call stack to find a file that has a different name than this one
+func getParentFilename() string {
+	_, currentFilename, _, ok := runtime.Caller(0)
+	if !ok {
+		return "Unknown"
+	}
+	for i := 1; i < 10; i++ {
+		_, parentFilename, _, ok := runtime.Caller(i)
+		if !ok {
+			return "Unkown"
+		} else if parentFilename != currentFilename {
+			return parentFilename
+		}
+	}
+	return "Unknown"
+}
+
+// creates a JSON representation of a log message
+func serializeLogMessage(ctx context.Context, message string) string {
```

Comment:
This just serializes the message component of the log object. A full log object (or line in a log file if it's easier to think of it that way) also contains the log level, a timestamp, etc. Its serialization format is dictated by `NewJSONLogWriter(...)` in `log4go_json_writer.go`. 

The only downside to this approach is that `context` and `logger` are sub-objects of the `message` component of the log line, as opposed to being top-level elements of the log object. We can't do much about this though, because l4g doesn't let us add arbitrary objects to a log line.